### DISCUSSION TITLE
Receive events of reading NFC from axis A8207

### DISCUSSION THREAD

**Discussion by ybloch** [AxisCommunications/discussions#23] [VAPIX]
I have a service, which manages entrances and exits.
I am interested in receiving direct card read events for my service by API HTTP PUSH from the AXIS reader.
Is there a way to do it? If so, how?

From the official support I received an answer to use pulling and make a request to the list of events of `/vapix/idpoint`, I tried it but I get 404 from almost every existing endpoint.

**Comment by pataxis** [comment]
Hi @ybloch,
thanks for the question, I will see if I can get hold of someone working with this product.

**Reply by ybloch** [reply]
Thanks, Looking forward to your results

**Comment by Corallo** [comment]
Hello @ybloch 

I have been in contact with the team that works closer to this type of device.
We recommend taking a look at the documentation [here](https://www.axis.com/vapix-library/subjects/t10167036/section/t10160259/display?section=t10160259-t10163438) in particular at the Websocket functions and Notification topics sections. There are examples that explain how to do what you are trying to achieve.

Let us know if you still have issues

**Reply by ybloch** [reply]
Hello @Corallo,
I was able to perform the actions listed there and subscribe to notifications through the API (`axev:subscribe`),
What I haven't understood yet, is where the messages will go. 
I didn't see a place where I set the IP of my server.
Can you please elaborate more on how the WS mod works? Who is the server and who is the client? Where does it send the events?

**Reply by Corallo** [reply]
The client is the application that establish the connection and subscribe to the event, in this case the "client" is your server who should subscribe to the event notification of the axis device.
Once your client subscribed the event, the Axis device will send a message to the client (yours server) when there is a triggering event.
In the link that I sent before, you can look for "Third party integration" there you can see the "handle_websocket_message" which gives an example of how the message sent by the device can be handled by your third party application.

**Reply by ybloch** [reply]
thanks,
I get it now,
But now I get 401 when I try to connect to WS, 
I tried to add the Authorization header by using some digest header builder, but without success
Here is my code:

```
from requests.auth import HTTPDigestAuth
import websocket
import rel

# creating auth header
chal = {'realm': 'AXIS_B8A44F4AFB81',
        'nonce': '1389832695:d3c620a9e645420228c5c7da7d228f8c'}
a = HTTPDigestAuth("username", "password")
a.init_per_thread_state()
a._thread_local.chal = chal
digest_header = a.build_digest_header('GET', '/vapix/intercomws')

# digest_header is now something like this: Digest username="username", realm="AXIS_B8A44F4AFB81", nonce="1389832695:d3c620a9e645420228c5c7da7d228f8c", uri="/vapix/intercomws", response="d5246a65eb0bdb83c6e******"

# connecting to WS
websocket.enableTrace(True)
ws = websocket.WebSocketApp("ws://192.168.1.123/vapix/intercomws",
                              on_open=on_open,
                              on_message=on_message,
                              on_error=on_error,
                              on_close=on_close,
                              header={"Authorization": digest_header})

# Set dispatcher to automatic reconnection
ws.run_forever(dispatcher=rel)
rel.signal(2, rel.abort)  # Keyboard Interrupt
rel.dispatch()
```

The response I get from the intercom is:
```
HTTP/1.1 401 Unauthorized
Date: Thu, 04 Aug 2022 07:53:00 GMT
Server: Apache/2.4.54 (Unix) OpenSSL/1.1.1q
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
WWW-Authenticate: Digest realm="AXIS_B8A44F4AFB81", nonce="ZU51o2XlBQA=7689bfdad2fd59827b2087020571d86343f50ffa", algorithm=MD5, stale=true, qop="auth"
Content-Length: 381
Content-Type: text/html; charset=iso-8859-1
-----------------------
Handshake status 401 Unauthorized
```

**Reply by Corallo** [reply]
First of all, make sure that the user "username" that you mention in your code really exist in your device.
Then, in the Digest access authentication you should first send a request, where you will get a response with 401 Unauthorized.
You should collect that response and extract the realm and nonce and use those in your auth header for future requests.

**Reply by ybloch** [reply]
Hi @Corallo  ,
I tried what you said without success, still getting 401, (the user exists in the Device).

Here is my code:

```
from requests.auth import HTTPDigestAuth
import websockets
import asyncio


def replace_nonce(req_header, resp_header):
    lines_without_nonce = [
        l for l in req_header.split(", ") if not "nonce" in l]
    lines_with_nonce = [l for l in resp_header.split(", ") if "nonce" in l]
    lines_without_nonce.append(lines_with_nonce[0])
    return ", ".join(lines_without_nonce)


async def connect():
    chal = {'realm': 'AXIS_B8A44F4AFB81',
            'nonce': '1389832695:d3c620a9e645420228c5c7da7d228f8c'}
    digest_builder = HTTPDigestAuth('root', "*******")
    digest_builder.init_per_thread_state()
    digest_builder._thread_local.chal = chal
    digest_header = digest_builder.build_digest_header(
        'GET', '/vapix/intercomws')
    print(digest_header)
    # OUTPUT:  Digest username="root", realm="AXIS_B8A44F4AFB81", nonce="1389832695:d3c620a9e645420228c5c7da7d228f8c", uri="/vapix/intercomws", response="d5246a65eb0bdb83c6ec44ed8c0b84bb"
    try:
        # The next line throws an exception: 401 Unauthorized
        async with websockets.connect("ws://192.168.1.149/vapix/intercomws", extra_headers={"Authorization": digest_header}) as websocket:
            response = await websocket.recv()
            print(response)
    except Exception as e:
        resp_auth = e.headers.get("www-authenticate")
        # resp_auth is: Digest realm="AXIS_B8A44F4AFB81", nonce="KdUZ0t/lBQA=7ffddd63b13ff150b8d90af4db8df06c8aae0fea", algorithm=MD5, stale=true, qop="auth"
        new_req_header = replace_nonce(digest_header, resp_auth)
        # new_req_header is: Digest username="root", realm="AXIS_B8A44F4AFB81", uri="/vapix/intercomws", response="d5246a65eb0bdb83c6ec44ed8c0b84bb", nonce="KdUZ0t/lBQA=7ffddd63b13ff150b8d90af4db8df06c8aae0fea"

        # The next line throws an exception: 401 Unauthorized
        async with websockets.connect("ws://192.168.1.149/vapix/intercomws", extra_headers={"Authorization": new_req_header}) as websocket:
            response = await websocket.recv()
            print(response)


async def main():
    await connect()

if __name__ == "__main__":
    asyncio.run(main())

```

If you could please send me a working piece of code, which connects to WS and receives RFID alerts with the TAG ID, it would save us a lot of precious time...

**Comment by vivekatoffice** [comment]
Hi @ybloch ,
You can follow [263](https://github.com/orgs/AxisCommunications/discussions/263) and try.
